test/powershell/engine/ParameterBinding/ParameterBinding.Tests.ps1

```diff
@@ -424,3 +424,331 @@
         }
     }
 }
+
+
+Describe "Custom type conversion in parameter binding" {
+    BeforeAll {
+        ## Prepare the script module
+        $content = @'
+    function Test-ScriptCmdlet {
+        [CmdletBinding(DefaultParameterSetName = "File")]
+        param(
+            [Parameter(Mandatory, ParameterSetName = "File")]
+            [System.IO.FileInfo] $File,
+
+            [Parameter(Mandatory, ParameterSetName = "StartInfo")]
+            [System.Diagnostics.ProcessStartInfo] $StartInfo
+        )
+
+        if ($PSCmdlet.ParameterSetName -eq "File") {
+            $File.Name
+        } else {
+            $StartInfo.FileName
+        }
+    }
+
+    function Test-ScriptFunction {
+        param(
+            [System.IO.FileInfo] $File,
+            [System.Diagnostics.ProcessStartInfo] $StartInfo
+        )
+
+        if ($null -ne $File) {
+            $File.Name
+        }
+        if ($null -ne $StartInfo) {
+            $StartInfo.FileName
+        }
+    }
+'@
+        Set-Content -Path $TestDrive\module.psm1 -Value $content -Force
+
+        ## Prepare the C# module
+        $code = @'
+    using System.IO;
+    using System.Diagnostics;
+    using System.Management.Automation;
+
+    namespace Test
+    {
+        [Cmdlet("Test", "BinaryCmdlet", DefaultParameterSetName = "File")]
+        public class TestCmdletCommand : PSCmdlet
+        {
+            [Parameter(Mandatory = true, ParameterSetName = "File")]
+            public FileInfo File { get; set; }
+
+            [Parameter(Mandatory = true, ParameterSetName = "StartInfo")]
+            public ProcessStartInfo StartInfo { get; set; }
+
+            protected override void ProcessRecord()
+            {
+                if (this.ParameterSetName == "File")
+                {
+                    WriteObject(File.Name);
+                }
+                else
+                {
+                    WriteObject(StartInfo.FileName);
+                }
+            }
+        }
+    }
+'@
+        $asmFile = [System.IO.Path]::GetTempFileName() + ".dll"
```

Comment:
Should you use $TestDrive here to ensure the assembly file gets cleaned up?

Reply:
Once the DLL is loaded, it will stay as long as powershell process is alive, and thus removing the file will throw errors. That's why I didn't use `$TestDrive`, so that Pester won't throw errors when trying to clean things up.